aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp

```diff
@@ -127,7 +133,112 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
+        "quantized::conv_prepack (qnnpack): Weights are expected to have 4 dimensions");
+    TORCH_CHECK(
+        stride.size() == 2, "quantized::conv_prepack (qnnpack): 2D convolution only");
+    TORCH_CHECK(
+        padding.size() == 2,
+        "quantized::conv_prepack (qnnpack): Specify top/left padding only. \
+       bottom/right padding assumed to be equal to top/left");
+    TORCH_CHECK(
+        dilation.size() == 2, " quantized::conv_prepack (qnnpack): 2D convolution only");
+
+    initQNNPACK();
+
+    // QNNPACK expects weights to be of the format {out_c, kH, kW, in_c/groups}
+    const size_t out_ch = weight.size(0);
+    const uint32_t kernel_h = weight.size(1);
+    const uint32_t kernel_w = weight.size(2);
+    const size_t in_ch = weight.size(3) * groups;
+
+    Tensor bias;
+    if (bias_in.has_value()) {
+      bias = bias_in.value();
+    } else {
+      bias = at::empty(out_ch, at::kFloat);
+      bias = at::quantize_linear(bias, 1.0, 0, kQInt32);
+    }
+    TORCH_CHECK(
+        !bias.defined() || (bias.ndimension() == 1 && bias.size(0) == out_ch),
+        "quantized::conv_prepack (qnnpack): expected bias to be 1-dimensional with ",
+        out_ch,
+        " elements",
+        ", but got bias of size ",
+        bias.sizes(),
+        " instead");
+
+    uint32_t stride_h = stride[0];
+    uint32_t stride_w = stride[1];
+    uint32_t pad_t = padding[0];
+    uint32_t pad_l = padding[1];
+    uint32_t dilation_h = dilation[0];
+    uint32_t dilation_w = dilation[1];
+
+    qnnpack::conv_param_t conv_p(
+        {kernel_w, kernel_h},
+        {stride_w, stride_h},
+        {dilation_w, dilation_h},
+        {pad_t, pad_l, pad_t, pad_l},
+        groups,
+        in_ch,
+        out_ch,
+        weight.q_zero_point(),
+        weight.q_scale(),
+        std::numeric_limits<uint8_t>::min(),
+        std::numeric_limits<uint8_t>::max());
+
+    auto weight_contig = weight.contiguous();
+    auto bias_contig = bias.contiguous();
+    auto wt_ptr = guts::make_unique<PackedConvWeightsQnnp>(
+        PackedConvWeightsQnnp{guts::make_unique<qnnpack::PrePackConvWeights>(
+                              conv_p,
+                              (uint8_t*)weight_contig.data_ptr<c10::quint8>(),
+                              (int32_t*)bias_contig.data_ptr<c10::qint32>()),
+                          {kernel_h, kernel_w},
+                          weight.q_scale(),
+                          weight.q_zero_point()});
+
+    return cpp_custom_type_hack::create(std::move(wt_ptr), weight.options());
+  }
+#endif // USE_PYTORCH_QNNPACK
+#if defined(USE_FBGEMM) || defined(USE_PYTORCH_QNNPACK)
+  Tensor operator()(
+      Tensor weight,
+      c10::optional<Tensor> bias,
+      torch::List<int64_t> stride,
+      torch::List<int64_t> padding,
+      torch::List<int64_t> dilation,
+      int64_t groups) {
+    auto& ctx = at::globalContext();
+#ifdef USE_FBGEMM
+    if (ctx.preferredQuantizedEngine() == at::QEngine::FBGEMM) {
```

Comment:
what is the default for the preferred engine? is it going to be set correctly in the mobile build (based on the compile flags)

Reply:
Default is currently set to FBGEMM. I can set the default for mobile build in a separate PR. Right now I'm setting it using `torch.backends.quantized.engine = torch.qnnpack` in the tests.